compact_str/src/repr2/mod.rs

```diff
@@ -0,0 +1,856 @@
+            }
+        } else {
+            let pointer = &self as *const _ as *const u8;
+            let length = core::cmp::min((last_byte.wrapping_sub(LENGTH_MASK)) as usize, MAX_SIZE);
+
+            // SAFETY: We know pointer is valid for `length` bytes
+            let slice = unsafe { core::slice::from_raw_parts(pointer, length) };
+            // SAFETY: A `Repr` contains valid UTF-8
+            let s = unsafe { core::str::from_utf8_unchecked(slice) };
+
+            String::from(s)
+        }
+    }
+
+    #[inline]
+    pub fn from_box_str(s: Box<str>) -> Self {
+        let og_cap = s.len();
+        let cap = Capacity::new(og_cap);
+
+        #[cold]
+        fn capacity_on_heap(s: Box<str>) -> Repr {
+            let heap = HeapBuffer::new(&s);
+            // SAFETY: `BoxString` and `Repr` are the same size
+            unsafe { mem::transmute(heap) }
+        }
+
+        #[cold]
+        fn empty() -> Repr {
+            EMPTY
+        }
+
+        if cap.is_heap() {
+            // We only hit this case if the provided String is > 16MB and we're on a 32-bit arch. We
+            // expect it to be unlikely, thus we hint that to the compiler
+            capacity_on_heap(s)
+        } else if og_cap == 0 {
+            // We don't expect converting from an empty String often, so we make this code path cold
+            empty()
+        } else {
+            // Don't drop the box here
+            let raw_ptr = Box::into_raw(s).cast::<u8>();
+            let ptr = ptr::NonNull::new(raw_ptr).expect("string with capacity has null ptr?");
+
+            // create a new BoxString with our parts!
+            let heap = HeapBuffer {
+                ptr,
+                len: og_cap,
+                cap,
+            };
+
+            // SAFETY: `BoxString` and `Repr` are the same size
+            unsafe { mem::transmute(heap) }
+        }
+    }
+
+    /// Reserves at least `additional` bytes. If there is already enough capacity to store
+    /// `additional` bytes this is a no-op
+    #[inline]
+    pub fn reserve(&mut self, additional: usize) {
+        let len = self.len();
+        let needed_capacity = len
+            .checked_add(additional)
+            .expect("Attempted to reserve more than 'usize' bytes");
+
+        if needed_capacity < self.capacity() {
+            // we already have enough space, no-op
+            return;
+        }
+
+        if needed_capacity <= MAX_SIZE {
+            // It's possible to have a `Repr` that is heap allocated with a capacity less than
+            // MAX_SIZE, if that `Repr` was created From a String or Box<str>
+            //
+            // SAFTEY: Our needed_capacity is >= our length, which is <= than MAX_SIZE
+            let inline = unsafe { InlineBuffer::new(self.as_str()) };
+            // SAFETY: `InlineBuffer` and `Repr` are the same size
+            *self = unsafe { mem::transmute(inline) };
```

Comment:
How about implementing `impl From<InlineBuffer> for Repr` and `impl From<HeapBuffer> for Repr` so that we don't need these `unsafe` transmute blocks everywhere?

Reply:
For now I'm going to stick with this pattern because I believe it'll be more performant. Otherwise we'll need to call through `fn from(...)` which would incur a performance penalty